[JunLong] iP

JunLongling opened the pull request:
# **Buu** 
## Buu frees your mind of having to remember things you need to do. 

Buu is a task management application that helps you organize tasks through a simple command-line interface. It supports the addition of to-do tasks, deadlines, and events, while also allowing you to mark, unmark, delete tasks, and view all tasks in a list.
> "_A simple and effective way to manage your tasks without clutter._"


## Key Features: ##

+ Task storage with easy loading and saving to a file
+ Supports different types of tasks: ToDo, Deadline, and Event
+ Ability to mark tasks as done or undone

> [!IMPORTANT]
> ## Commands Supported: ##

1. `todo [task description]` - To add a ToDo task.
2. `deadline [task description] /by [date/time]` - To add a Deadline task.
3. `event [task description] /from [start date/time] /to [end date/time]` - To add an Event task.
4. `list` - To display all tasks.
5. `mark [task number]` - To mark a task as done.
6. `unmark [task number]` - To unmark a task.
7. `delete [task number]` - To delete a task.
8. `bye` - To exit the chatbot.

``` ruby
public Buu() {
        String filePath = "data/tasks.txt"; // Default file path
        ui = new Ui();
        storage = new Storage(filePath);
        taskList = new TaskList(storage.loadTasks());
        // Show the welcome message when the Buu application starts
        ui.showWelcomeMessage("Buu"); // This adds the welcome message
    }
```

## Task List: ##
- [x] Implement abstract Task class (done)
- [x] Create ToDo, Deadline, and Event subclasses (done)
- [ ] Complete the GUI (not done)

For more details, please click [here](https://github.com/nus-cs2103-AY2425S1/ip/pull/197)
Use todo, deadline, or event to start organizing your tasks today. ✨

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
Completed Level 0. Rename, Greet, Exit

Commit:
completed Level 1. Echo

Commit:
Completed Level 2. Add, List

Commit:
Completed Level 3. Mark as Done

Commit:
Completed Level 4. ToDos, Events, Deadlines

Commit:
completed  Duke A-TextUiTesting: Automated Text UI Testing

Commit:
completed Level 5. Handle Errors

Commit:
completed Level 6. Delete

Commit:
Completed A-Enums: Use Enums

Commit:
Add Save

Commit:
add deadline

Commit:
add deadline

Commit:
add level 8

Commit:
Add A-MoreOOP

Commit:
Add A-Packages

Commit:
Add A-Gradle

Commit:
Add A-JUnit

Commit:
Add A-Jar

Commit:
Add A-JavaDoc

Commit:
Add Level-9

Commit:
Add A-JavaDoc

cth06-Github commented on build.gradle:
```diff
@@ -0,0 +1,33 @@
+    archiveClassifier = null
+}
+
+run{
```
```suggestion
run {
```

JunLongling replied:
Thank you for pointing this out! I see that the spacing for the opening brace { was inconsistent. I’ve updated the code accordingly

cth06-Github commented on src/main/java/GPT/Deadline.java:
```diff
@@ -0,0 +1,56 @@
+     * Constructs a Deadline task with the specified description and deadline.
+     *
+     * @param description The description of the task.
+     * @param by          The date and time by which the task should be completed.
```
Nice! I see that you are trying to align the additional information for the parameter "by" with the rest, resulting in a many white spaces between parameter "by" and the additional information! I think it is a really nice way of organising though perhaps it's not necessary to be in that precise alignment :) (I think there is no right or wrong answer here).

JunLongling replied:
Thank you for the positive feedback! I appreciate your suggestion regarding the alignment. I initially aligned the parameters for consistency and readability, but I agree that it might not be necessary to be this precise. I will simplify the spacing to make it less rigid while keeping the documentation clear.

cth06-Github commented on src/main/java/GPT/DeadlineCommand.java:
```diff
@@ -0,0 +1,56 @@
+    @Override
+    public void execute(TaskList taskList, Ui ui, Storage storage) {
+        if (description == null || by == null) {
+            ui.showError("Invalid command format for deadline.\nUsage: deadline [task description] /by [date/time]\nExample: deadline return book /by 2/12/2019 1800");
```
Would you wish to break this line of code (around 154 characters) into 2 lines?

JunLongling replied:
Thank you for pointing this out! I have broken the line into two, as suggested, to ensure it complies with the coding standard.

cth06-Github commented on src/main/java/GPT/DeleteCommand.java:
```diff
@@ -0,0 +1,39 @@
+            taskList.deleteTask(index);
+            storage.saveTasks(taskList.getTasks());
+            ui.showTaskRemoved(removedTask, taskList.getTasks().size());
+        } catch (GPTException e) {
```
Would gptException be the correct naming standards?

JunLongling replied:
Thank you for pointing that out! I agree that gptException would follow the correct camelCase naming convention for the exception variable. I will make the change for consistency.

cth06-Github commented on src/main/java/GPT/Event.java:
```diff
@@ -0,0 +1,61 @@
+    @Override
+    public String toString() {
+        if (from != null && to != null) {
+            return super.toString() + " (from: " + from.format(OUTPUT_FORMATTER) + " to: " + to.format(OUTPUT_FORMATTER) + ")";
```
I think this code of line can be broken into 2 lines!

cth06-Github commented on src/main/java/GPT/EventCommand.java:
```diff
@@ -0,0 +1,62 @@
+     */
+    public EventCommand(String input) {
+        String[] parts = input.substring(5).trim().split("/from|/to");
+        if (parts.length < 3 || parts[0].trim().isEmpty() || parts[1].trim().isEmpty() || parts[2].trim().isEmpty()) {
```
I think this code of line can be separated into 2 lines!
I think this phenomenon (code of line should be split as the number of characters exceed 110)  can be found in other files as well for you to consider whether to rectify them!

JunLongling replied:
Thank you for pointing this out! I have gone through the codebase and resolved all similar issues where the lines exceeded the character limit. The lines have been split accordingly for better readability and to follow the coding standards.

Commit:
Add A-CheckStyle

cth06-Github commented on src/test/java/GPT/EventTest.java:
```diff
@@ -0,0 +1,47 @@
+package GPT;
+
+import org.junit.jupiter.api.Test;
```
Would it better to break a new line here? To keep it consistent with [src/test/java/GPT/DeadlineTest.java](https://github.com/nus-cs2103-AY2425S1/ip/pull/197/files#diff-208c4ea4e8815989c3c5aea536ec0e1460974c8368bfc50fea4c2b91aa976f21). (or vice versa).

JunLongling replied:
Thank you for the suggestion! I’ve updated the code and fixed the line break for consistency

cth06-Github reviewed (commented):
Overall, the coding standards are mostly adhered (or at least, I can't detect much violations), good job!

Commit:
fix: address reviewer feedback for consistency in line breaks

Commit:
Add Level 10. GUI

Commit:
no message

Commit:
Review for varargs applicability; no changes needed

darkmoongreatsword commented on src/main/java/Buu/Buu.java:
```diff
@@ -0,0 +1,68 @@
+        storage = new Storage(filePath);
+        taskList = new TaskList(storage.loadTasks());
+        // Show the welcome message when the Buu application starts
+        ui.showWelcomeMessage("Buu"); // This adds the welcome message
```
It is generally not good practice for a constructor to have side-effects such as displaying UI elements. It might be better to move this down to an initialisation method such as in main() itself.

darkmoongreatsword commented on src/main/java/Buu/Buu.java:
```diff
@@ -0,0 +1,68 @@
+            return response.toString().trim(); // Return the full output to display in the GUI
+        } catch (GPTException e) {
+            return "Error: " + e.getMessage();
+        } catch (Exception e) {
```
Catching the general Exception here may obscure the source of specific, more expected errors. It might be better to break it down to handle more specific errors like IOError.

darkmoongreatsword commented on src/main/java/Buu/Ui.java:
```diff
@@ -0,0 +1,150 @@
+    public void addToOutputBuffer(String message) {
+        outputBuffer.add(message);
+    }
+}
```
Very good application of OOP principles, with separation into modular components. The methods are very readable, and intuitive. It will be easy to extend this class further for more features involving UI.

darkmoongreatsword commented on src/main/java/Buu/Parser.java:
```diff
@@ -0,0 +1,86 @@
+     * @throws GPTException If the input cannot be parsed into a valid command.
+     */
+
+    public static Command parseCommand(String input) throws GPTException {
```
A switch-case may be more preferable to this long line of if-else statements.

darkmoongreatsword reviewed (commented):
Looks good.
Overall, very well written code, with clear documentation.
There are some minor violations in code quality and standards like
catching very general Exceptions.
However, it does not affect the readability significantly.

Commit:
Update build.gradle to generate a new JAR file

The current configuration does not generate a JAR file, which
is necessary for distribution.

This change adds the necessary tasks to build.gradle for creating
a JAR file during the build process.

Commit:
Implement assert feature to document important assumptions in code

The code now includes assertions to check for critical assumptions at various points. These assertions validate important preconditions, such as ensuring non-null inputs, valid indices, and properly initialized objects. Additionally, assertions document control-flow and class invariants, such as ensuring that certain values remain within expected ranges.

This change enhances the robustness of the code by helping catch potential issues early during development. Assertions are not intended for normal error handling and will be disabled in production builds for performance reasons.

JunLongling committed:
Merge pull request #1 from JunLongling/branch-A-Assertions

Add assertions to check critical assumptions

Commit:
Refactor Parser class to eliminate arrowhead anti-pattern

Refactored the `parseDateTime` method in the `Parser` class to remove
nested try-catch blocks and avoid the arrowhead anti-pattern.
This improves code readability and maintainability by reducing
deep nesting and simplifying the control flow.

The refactoring maintains the functionality of parsing date strings
using multiple predefined formats, ensuring invalid inputs are
handled appropriately and providing feedback for invalid formats.

JunLongling committed:
Merge branch 'master' into branch-A-CodeQuality

JunLongling committed:
Merge pull request #2 from JunLongling/branch-A-CodeQuality

Refactor Parser class to eliminate arrowhead anti-pattern

Commit:
Refactor Parser class to use Streams for date parsing

Refactored the `parseDateTime` method in the `Parser` class to utilize
Java Streams for cleaner and more efficient code. Streams are now
used to attempt parsing with various date and time formats, and
Optional is returned to handle success or failure gracefully.

This approach reduces nested if-else statements (arrowhead code)
and enhances readability and maintainability of the code. DateTime
and Date parsing functions have been separated for better structure.

JunLongling committed:
Merge pull request #3 from JunLongling/branch-A-Streams

Refactor Parser class to use Streams for date parsing

Commit:
Set up Continuous Integration with GitHub Actions

This commit adds GitHub Actions for continuous integration. The workflow is
configured to run tests automatically on each push to the repository. This
ensures that code changes meet quality standards before being merged.

CI tasks include:
- Compiling the codebase.
- Running the unit tests.
- Reporting test results for feedback.

With this setup, the team will benefit from automatic validation of code
before it is merged, ensuring higher code quality and reducing potential
issues.

Commit:
Implement C-Priority for task management system
- Updated related classes and methods to integrate priority settings.
- Ensured correct priority display in task representations.
- Added test cases to verify priority handling.

Details:
- Implemented priority management in Deadline, Event, and ToDo tasks.
- Verified functionality with JUnit tests.

Commit:
Refactor Ui class to align with Buu's personality

The Ui class messages previously included references to tasks as candy,
which was inconsistent with the intended task-focused behavior.
Additionally, the chatbot's personality did not fully align with Buu's
playful and energetic character from the anime.

Refactor the Ui class to remove candy references and update messages to
better reflect Buu’s personality. This ensures that the chatbot remains
fun and engaging while staying focused on task management.

Commit:
This commit enhances the code to handle various anticipated errors
during product usage. Improved error handling will provide
users with clearer feedback and prevent unexpected crashes.

Handled errors include:
- Command format errors such as extra spaces, missing or duplicate
  parameters, and unexpected special characters.
- Environment issues such as missing or inaccessible files and invalid
  file content.
- Data validation errors like start date/time later than end date/time,
  duplicated values where uniqueness is expected, and invalid dates.

These improvements ensure more robust error handling, improving user
experience and reducing the likelihood of incorrect program behavior.

Commit:
SetPriorityCommand: ensure task index validation and throw exception

The SetPriorityCommand class did not throw an exception when
the task index was out of bounds, leading to test failures for
invalid task index scenarios.

This behavior caused an inconsistency when invalid input was provided,
as errors were displayed via the UI but not
properly handled in the code.

Let's throw an IllegalArgumentException when the task index
is out of bounds in the execute() method. This will ensure that
invalid input is properly handled and the corresponding
test cases pass.

This change improves the robustness of task index validation and ensures
better error handling for out-of-range task indices.

Commit:
Add final touch-up and ensure all functions are tested thoroughly

This commit finalizes the testing by adding comprehensive test coverage
for all functions in the project. The test cases ensure correct
behavior for edge cases, valid inputs, and invalid inputs across
all commands and functions.

Key changes include:
- Added missing test cases for boundary conditions.
- Enhanced existing test cases to handle edge cases.
- Verified that error handling works correctly for all scenarios.
- Ensured that the task index validation and priority setting logic are
  thoroughly tested, including out-of-bounds inputs.

These changes guarantee that all functions are working correctly
and improve the robustness and reliability of the project.

Commit:
Add README.md and Ui.png for project documentation

This commit adds the README.md file that provides an overview of the
Buu Task Manager project, including features, usage instructions,
project structure, and contribution guidelines.

Additionally, the Ui.png screenshot is included to visually enhance
the README documentation, giving users a better understanding of
the application's interface.

These files improve the clarity of the project and provide necessary
documentation for users and contributors.

JunLongling committed:
Update README.md

Commit:
Refactor task classes and tests:

- Updated Event and Deadline classes to use LocalDateTime
instead of String for date/time handling.
- Modified toFileFormat and toString methods for proper date formatting.
- Enhanced unit tests for Event and Deadline to use LocalDateTime,
ensuring correct task creation and formatting.
- Improved Storage class to correctly handle task loading
and saving with updated formats."

Commit:
Enhance command classes for user-friendly error handling

Updated MarkCommand, DeleteCommand, and UnmarkCommand to utilize the
Ui class for displaying error messages. Improved input validation
and removed unnecessary exception handling in constructors.

JunLongling committed:
Update README.md

Revised the README to include detailed descriptions of commands, 
examples of usage, and clearer instructions for new users. 
Added sections on installation and features to enhance user 
understanding of the application.

JunLongling committed:
Update README.md

JunLongling committed:
Update README.md
